pkg/kubectl/cmd/set/set_env.go

```diff
@@ -173,18 +181,66 @@ func keyToEnvName(key string) string {
 	return strings.ToUpper(validEnvNameRegexp.ReplaceAllString(key, "_"))
 }
 
-func (o *EnvOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
-	if o.All && len(o.Selector) > 0 {
-		return fmt.Errorf("cannot set --all and --selector at the same time")
+func contains(key string, keyList []string) bool {
+	if len(keyList) == 0 {
+		return true
 	}
-	resources, envArgs, ok := envutil.SplitEnvironmentFromResources(args)
-	if !ok {
-		return cmdutil.UsageErrorf(o.Cmd, "all resources must be specified before environment changes: %s", strings.Join(args, " "))
+
+	for _, k := range keyList {
+		if k == key {
+			return true
+		}
 	}
-	if len(o.Filenames) == 0 && len(resources) < 1 {
-		return cmdutil.UsageErrorf(cmd, "one or more resources must be specified as <resource> <name> or <resource>/<name>")
+	return false
+}
+
+func (o *EnvOptions) getConfigResources(f cmdutil.Factory, cmd *cobra.Command) ([]*v1.ConfigMap, []*v1.Secret, error) {
+	if len(o.From) == 0 {
+		return nil, nil, nil
+	}
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return nil, nil, err
 	}
+	b := f.NewBuilder().
+		Internal().
+		LocalParam(o.Local).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, &o.FilenameOptions).
+		Flatten()
+
+	if !o.Local {
+		b = b.
+			LabelSelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, o.From).
+			Latest()
+	}
+
+	infos, err := b.Do().Infos()
+	if err != nil {
+		return nil, nil, err
+	}
+	configMaps := []*v1.ConfigMap{}
+	secrets := []*v1.Secret{}
+	for _, info := range infos {
+		versionedObject, err := info.Mapping.ConvertToVersion(info.Object, info.Mapping.GroupVersionKind.GroupVersion())
+		if err != nil {
+			return nil, nil, err
+		}
+		switch from := versionedObject.(type) {
+		case *v1.Secret:
+			secrets = append(secrets, from)
+		case *v1.ConfigMap:
+			configMaps = append(configMaps, from)
+		default:
+			return nil, nil, fmt.Errorf("unsupported resource specified in --from")
```

Comment:
maybe it'd help a user if we included which resources are supported in this error